### ISSUE TITLE
Need to extend ImageBitmap clearly to wrap MediaStreamTrack

### ISSUE THREAD

**Issue by ds-hwang** [ChiahungTai/mediacapture-worker#2]
[ImageBitmap spec](https://html.spec.whatwg.org/multipage/webappapis.html#images) declares that `HTMLVideoElement` is one of `ImageBitmapSource`. When `Window.createImageBitmap(HTMLVideoElement)` is called, `ImageBitmap` captures the frame at the moment.

This spec says `The input video frame comes from the corresponding MediaStreamTrack.` but there is not explanation how `MediaStreamTrack` is corresponding `ImageBitmap`. In my opinion, `MediaStreamTrack` should be one of `ImageBitmapSource` like `HTMLVideoElement` if this spec want to use `ImageBitmap` as `inputImageBitmap`. However, role of `MediaStreamTrack` is a bit duplicated to `HTMLVideoElement`. How this issue is resolved is important to [mediacapture-depth](https://github.com/w3c/mediacapture-depth/) @huningxin , @anssiko 

In addition, it is ambiguous what `outputImageBitmap` refers to among `ImageBitmapSource` under the hood. If it is `ImageData` or `Blob`, it's very inefficient to draw it on WebGL or Canvas from GPU perspective.

**Comment by ds-hwang** [comment]
After reading [FoxEye slide 22p](https://wiki.mozilla.org/images/1/1f/Project_FoxEye_2015-Feb.pdf), I understand why `outputImageBitmap` is introduced. I disagree on `outputImageBitmap`. In theory, there is no reason that `canvas.putImageData()` is slower than `canvas.drawImage(ImageBitmap)`. I think it's FireFox bug. 
`outputImageBitmap` is unnecessary. IMO, how to pass output from worker to main thread is up to web developers. They may use `ArrayBuffer` or `ImageData`.

**Comment by ChiahungTai** [comment]
Sorry for the misunderstanding. I add more descriptions in VideoProcessEvent section. Basically the key point is we don't use  `Window.createImageBitmap(HTMLVideoElement)`. The implementation will let the `MediaStreamTrack` dispatches every unique frame to the `VideoWorker`s which are added to the `MediaStreamTrack`. You can take this as the `MediaStreamTrack` will create multiple `ImageBitmap`s and send to the event handler `onvideoprocess`. But if you take `HTMLVideoElement` as the source of `ImageBitmap`, it only create only one `ImageBitmap`. I think the main different is push v.s pull mechanism and worker model. Just like `FrameGrabber` instead of `CanvasRender2DContext`.

The FoxEye slide is needed to be updated. We will provide the latest slide in end of June. Only WorkerProcessor cases need `outputImageBitmap`. The web developers don't need `outputImageBitmap` in WorkerMonitor. `addWorkerProcessor` will generate a new `MediaStreamTrack`. The input source of this new generated `MediaStreamTrack` is the `outputImageBitmap` in the `VideoProcessEvent`. That is the reason why we need `outputImageBitmap`. Actually the web developers can create a processed `MediaStreamTrack` chain by `addWorkerProcessor`. Hope this comment would answer your question.

**Comment by ds-hwang** [comment]
Does it mean that your spec extends `MediaStreamTrack` is one of `ImageBitmapSource` ?
Could you elaborate more about "You can take this as the MediaStreamTrack will create multiple ImageBitmaps"? As I understand, you want to create multiple `ImageBitmap`s for multiple video frames coincidentally.

**Comment by ChiahungTai** [comment]
No, `MediaStreamTrack` is not one of `ImageBitmapSource`. Every frame of `MediaStreamTrack` will belongs to one InputImageBitmap. Create multiple `ImageBitmap` means `MediaStreamTrack` should have streamming video frame. Not like createImageBitmap only takes one video frame.

**Comment by ds-hwang** [comment]
ImageBitmap is introduced to snapshot video, canvas, or webgl. I think you are against the purpose of ImageBitmap.
From implementation perspective, it's ambiguous that MediaStreamTrack belongs to ImageBitmap.

**Comment by ChiahungTai** [comment]
No, I mean every video frame of MST(after the association between MST and VideoWorker created). Not MST itself. I think ervery video frame is equal to every snapshot of MST. Feel free to ask.more question. Your feedback is good for me to pollish the spec.

**Comment by ds-hwang** [comment]
every video frame of VideoElement belongs to ImageBitmap. It's same to MST. If you want to create ImageBitmap including every video frame of MST, in my opinion, you have to define MST is one of ImageBitmapSource like VideoElement.
Extending window.createImageBitmap(MST) should be prerequisite of video worker.

**Comment by ChiahungTai** [comment]
Let me summarize all the questions and make sure I answer your question. Let's discuss whether `MediaStreamTrack` is one of `ImageBitmapSource` later. 
1. there is not explanation how MediaStreamTrack is corresponding ImageBitmap => Did I answer your question? I also add more description in `VideoProcessEvent` interface section.
2. However, role of MediaStreamTrack is a bit duplicated to HTMLVideoElement. =>Relative is `MediaStreamTrack` one of `ImageBitmapSource`? Discuss later.
3. We assume the `ImageBitmapSource` of `outputImageBitmap`  is 'ArrayBuffer'. Good point. We should consider it.
4. Do you still think `outputImageBitmap` is unnecessary after my explanation? 
5. Did you have any other question except the question "is `MediaStreamTrack` one of `ImageBitmapSource`?".

**Comment by ds-hwang** [comment]
sorry for late response.
1. Before defining that MediaStreamTrack is one of ImageBitmapSource, IMO it's still ambiguous how implementers `inputImageBitmap`.
2. ok, discuss later.
3. if `outputImageBitmap` wrapped `ArrayBuffer`, why `outputImageBitmap` is needed? transferring `ArrayBuffer` is more succinct.
4. unfortunately, yes.
5. yes, I have some suggestion how to support both worker and main thread in the spec. I'll file another issue.

**Comment by ChiahungTai** [comment]
hi, ds,
Thanks for your response.
About 3 and 4, can you give me an example to explain how to support UC6 in [Media_Capture_Depth_Stream_Extension#Examples](https://www.w3.org/wiki/Media_Capture_Depth_Stream_Extension#Examples) without `outputImageBitmap`?
Thanks,

**Comment by ds-hwang** [comment]
Hi Chiahung,
I rephrase EXAMPLE 8 in the spec,
processor.js in EXAMPLE 8:

```
onvideoprocess = function(event) {
  var uInt8Array = new Uint8Array(new ArrayBuffer(event.inputImageBitmap.size()));
  processOneFrame_Invert(event.inputImageBitmap, uInt8Array);
  self.postMessage({'outputImage': uInt8Array});
}
```

FYI, ArrayBuffer is [transferable objects](http://updates.html5rocks.com/2011/12/Transferable-Objects-Lightning-Fast), so it's zero-copy operation.

**Comment by ChiahungTai** [comment]
@ds-hwang, @kakukogou and I will answer all the questions in recent days. @kakugou is busy at landing ImageBitmap into gecko. It is almost done. He will update the spec of ImageBitmap extension. After that, that will be much easier to answer your questions. :)
